File: src/main/java/org/moinex/model/investment/FundamentalAnalysis.java
/*
 * Filename: FundamentalAnalysis.java
 * Created on: January  9, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.model.investment;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.moinex.model.enums.PeriodType;
import org.moinex.util.Constants;

/**
 * Entity that stores fundamental analysis data for a ticker
 * This data is fetched from the Python script and cached in the database
 * Linked to Ticker entity - analysis is only available for active (non-archived) tickers
 */
@Entity
@Table(name = "fundamental_analysis")
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Builder
public class FundamentalAnalysis {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @Setter(AccessLevel.NONE)
    private Long id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ticker_id", nullable = false, unique = true)
    private Ticker ticker;

    @Column(name = "company_name")
    private String companyName;

    @Column(name = "sector")
    private String sector;

    @Column(name = "industry")
    private String industry;

    @Column(name = "currency", nullable = false)
    private String currency;

    @Column(name = "period_type", nullable = false)
    private PeriodType periodType;

    @Column(name = "data_json", columnDefinition = "TEXT", nullable = false)
    private String dataJson; // Full JSON data from Python script

    @Column(name = "last_update", nullable = false)
    private String lastUpdate;

    @Column(name = "created_at", nullable = false)
    private String createdAt;

    public LocalDateTime getLastUpdate() {
        return LocalDateTime.parse(lastUpdate, Constants.DB_DATE_FORMATTER);
    }

    public void setLastUpdate(LocalDateTime lastUpdate) {
        this.lastUpdate = lastUpdate.format(Constants.DB_DATE_FORMATTER);
    }

    public LocalDateTime getCreatedAt() {
        return LocalDateTime.parse(createdAt, Constants.DB_DATE_FORMATTER);
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt.format(Constants.DB_DATE_FORMATTER);
    }
}

File: src/main/kotlin/org/moinex/model/investment/FundamentalAnalysis.kt
/*
 * Filename: FundamentalAnalysis.kt (original filename: FundamentalAnalysis.java)
 * Created on: January  9, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.model.investment

import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import org.moinex.common.converter.LocalDateTimeStringConverter
import org.moinex.model.enums.PeriodType
import java.time.LocalDateTime

@Entity
@Table(name = "fundamental_analysis")
class FundamentalAnalysis(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Int? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ticker_id", nullable = false)
    var ticker: Ticker,
    @Column(name = "company_name")
    var companyName: String? = null,
    @Column(name = "sector")
    var sector: String? = null,
    @Column(name = "industry")
    var industry: String? = null,
    @Column(name = "currency", nullable = false)
    var currency: String,
    @Enumerated(EnumType.STRING)
    @Column(name = "period_type", nullable = false)
    var periodType: PeriodType,
    @Column(name = "data_json", columnDefinition = "TEXT", nullable = false)
    var dataJson: String,
    @Column(name = "last_update", nullable = false)
    @Convert(converter = LocalDateTimeStringConverter::class)
    var lastUpdate: LocalDateTime,
    @Column(name = "created_at", nullable = false)
    @Convert(converter = LocalDateTimeStringConverter::class)
    var createdAt: LocalDateTime,
) {
    init {
        currency = currency.trim().uppercase()

        require(currency.isNotEmpty()) {
            "Currency cannot be empty"
        }
        require(dataJson.isNotEmpty()) {
            "Data JSON cannot be empty"
        }
    }

    override fun toString(): String = "FundamentalAnalysis [id=$id, ticker=${ticker.symbol}, period=$periodType]"
}
